### PR TITLE
Adding WebSocket APIs

### PR DESCRIPTION
**Pull request by jasonsandlin**

### CODE REVIEW

File: Source/WebSocket/websocket.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        return HC_E_INVALIDARG;
+    }
+
+    HC_TRACE_INFORMATION(WEBSOCKET, "HCWebSocketCloseHandle [ID %llu]", websocket->id);
+    int refCount = --websocket->refCount;
+    if (refCount <= 0)
+    {
+        assert(refCount == 0); // should only fire at 0
+        delete websocket;
+    }
```
Comment (jicailiu) [inline Source/WebSocket/websocket.cpp]: do we need to worry about thread safe?

File: Source/WebSocket/websocket.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+try 
+{
+    auto httpSingleton = get_http_singleton();
+
+    auto connectFunc = httpSingleton->m_websocketConnectFunc;
+    if (connectFunc != nullptr)
+    {
+        try
+        {
+            connectFunc(uri, websocket, args);
```
Comment (jicailiu) [inline Source/WebSocket/websocket.cpp]: >connectFunc [](start = 12, length = 11)

should set connectCalled to true

Comment (jasonsandlin) [reply Source/WebSocket/websocket.cpp]: Will do

File: Build/libHttpClient.141.UWP.C/libHttpClient.141.UWP.C.vcxproj [hunk 3]
```diff
@@ -118,9 +118,10 @@
   </PropertyGroup>
   <ItemGroup>
     <ClCompile Include="$(MSBuildThisFileDirectory)..\..\Source\Common\Win\utils_win.cpp" />
+    <ClCompile Include="..\..\Source\WebSocket\websocket.cpp" />
```
Comment (jasonsandlin) [inline Build/libHttpClient.141.UWP.C/libHttpClient.141.UWP.C.vcxproj]: add to cmake

File: Include/httpClient/httpClient.h [hunk 4]
```diff
@@ -274,13 +274,23 @@ HCHttpCallPerform(
...
+/// <summary>
+/// Decrements the reference count on the call object. 
+/// When the HC_CALL_HANDLE ref count is 0, HCHttpCallCloseHandle() will 
+/// free the memory associated with the HC_CALL_HANDLE
 /// </summary>
 /// <param name="call">The handle of the HTTP call</param>
 /// <returns>Result code for this API operation.  Possible values are HC_OK, HC_E_INVALIDARG, or HC_E_FAIL.</returns>
 HC_API HC_RESULT HC_CALLING_CONV
-HCHttpCallCleanup(
+HCHttpCallCloseHandle(
```
Comment (jasonsandlin) [inline Include/httpClient/httpClient.h]: fix breaks in other samples/tests

File: Include/httpClient/httpClient.h [hunk 5]
```diff
@@ -558,6 +547,159 @@ HCHttpCallResponseGetHeaderAtIndex(
...
+/// <summary>
+/// Connects to the WebSocket.  This API returns immediately and will spin up a thread under the covers.
+/// On UWP and XDK, the thread is owned and controlled by Windows::Networking::Sockets::MessageWebSocket
+/// </summary>
+/// <param name="uri">The URI to connect to</param>
+/// <param name="websocket">The handle of the WebSocket</param>
+/// <param name="args">Struct for describing the WebSocket connection args</param>
+/// <returns>Result code for this API operation.  Possible values are HC_OK, HC_E_INVALIDARG, HC_E_OUTOFMEMORY, or HC_E_FAIL.</returns>
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketConnect(
```
Comment (jicailiu) [inline Include/httpClient/httpClient.h]: >HCWebSocketConnect [](start = 0, length = 18)

Connect and send should be async, close is technically async, but we can fire and forget.

File: Source/WebSocket/websocket.cpp [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+    HC_TRACE_INFORMATION(WEBSOCKET, "HCWebSocketDuplicateHandle [ID %llu]", websocket->id);
+    ++websocket->refCount;
+
+    return websocket;
+}
+CATCH_RETURN_WITH(nullptr)
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketCloseHandle(
```
Comment (jicailiu) [inline Source/WebSocket/websocket.cpp]: >HCWebSocketCloseHandle [](start = 0, length = 22)

websocket close is talk to service and perform a close communication if possible. I don't think you can close multiple times.

File: Include/httpClient/httpClient.h (hunk 5 above)
Comment (jasonsandlin) [reply Include/httpClient/httpClient.h]: Updated

File: Build/libHttpClient.141.UWP.C/libHttpClient.141.UWP.C.vcxproj (hunk 3 above)
Comment (jasonsandlin) [reply Build/libHttpClient.141.UWP.C/libHttpClient.141.UWP.C.vcxproj]: Fixed

**Pull request merged**